### PR TITLE
Implemented changes to LSP versions 3.6 and 3.7

### LINKED ISSUE #167: Adapt to Protocol Level 3.6
https://github.com/Microsoft/language-server-protocol/blob/master/protocol-changelog.md

### PR DESCRIPTION
**Pull request by spoenemann**
Fixes #167, #178, and #118.

Unfortunately, this change massively breaks API. I don't know how else we could realize the change in LSP from

* method: ‘textDocument/completion’
* params: `TextDocumentPositionParams`

to

* method: ‘textDocument/completion’
* params: `CompletionParams`

In TypeScript this change is API compatible, in Java it's not :-(

### CODE REVIEW

File: org.eclipse.lsp4j/src/main/xtend-gen/org/eclipse/lsp4j/WorkspaceFolder.java [hunk 1]
```diff
@@ -26,6 +30,18 @@
    */
   private String name;
```
Comment (akosyakov) [inline org.eclipse.lsp4j/src/main/xtend-gen/org/eclipse/lsp4j/WorkspaceFolder.java]: In the spec it is not marked as optional. Is it bug there?

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 2]
```diff
@@ -1618,7 +1869,7 @@ class MarkupContent {
 	 * The type of the Markup.
 	 */
 	@NonNull
-	String MarkupKind;
+	String kind;
```
Comment (akosyakov) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: It is named as `kind` in the spec

**Comment by jvalkeal** [comment]
Just out of curiosity if you could start over, what comes for a `textDocument/completion` and other objects, would you simply go with `TextDocumentCompletion` and then postfixing request/response classes with `Request` or `Response`? Then all changes would be internal to those classes.

File: org.eclipse.lsp4j/src/main/xtend-gen/org/eclipse/lsp4j/WorkspaceFolder.java (hunk 1 above)
Comment (spoenemann) [reply org.eclipse.lsp4j/src/main/xtend-gen/org/eclipse/lsp4j/WorkspaceFolder.java]: The LSP documentation says _"The name of the workspace folder. Defaults to the uri's basename."_ I think it's a bug in the specification: either the property should be optional, or the second part of the property documentation should be removed.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 2 above)
Comment (spoenemann) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: That's why I corrected it here :-)

**Comment by spoenemann** [comment]
@akosyakov I'll prepare a change log for the release.

@jvalkeal I don't understand what you mean. Could you explain further?

**Comment by jonahkichwacoders** [comment]
While you are updating LS versions, it may be worth updating the Readme, it currently says:

### Supported LSP Versions

 * LSP4J 0.2.x &rarr; LSP 3.x
 * LSP4J 0.1.x &rarr; LSP 2.x

**Comment by spoenemann** [comment]
Up to now we have only implemented the data types from the specification, and have not added any additional types.

Even then we wouldn't get around breaking the API. The root of the problem is that the LSP is designed for TypeScript, which has a completely different type system than Java. Some kinds of API changes that are 100% compatible in TypeScript break the corresponding Java representations.

**Comment by spoenemann** [comment]
https://github.com/eclipse/lsp4j/blob/master/CHANGELOG.md

**Pull request merged**